File: express/src/routes/Tournaments/tournaments.ts
import { Router } from 'express';
import { prisma } from '../../lib/prisma.js';
import { TournamentService } from '../../services/Tournaments/tournament_service.js';

const router = Router();
const service = new TournamentService();

function validate(data: any): void {
  if (!((data.maxPlayers == null || (data.maxPlayers >= 2 && data.maxPlayers <= 512)))) throw new Error(`Tournament must allow between 2 and 512 players`);
  if (!((data.entryFee == null || Number(data.entryFee) >= 0))) throw new Error(`Entry fee must not be negative`);
  if (!((data.prizePool == null || Number(data.prizePool) >= 0))) throw new Error(`Prize pool must not be negative`);
  if ((data.endTime != null) && !((data.endTime == null || (data.startTime != null && data.endTime > data.startTime)))) throw new Error(`End time must be after start time`);
}
function applyProjection(obj: any): any {
  if (!obj) return obj;
  const r = { ...obj };
  if ('createdAt' in r) { r.createdAt = r.createdAt; delete r.createdAt; }
  if ('startTime' in r) { r.startTime = r.startTime; delete r.startTime; }
  if ('endTime' in r) { r.endTime = r.endTime; delete r.endTime; }
  return r;
}

const ALLOWED_TRANSITIONS: Record<string, string[]> = {
  'Draft': ['Registration'],
  'Registration': ['Ongoing', 'Cancelled'],
  'Ongoing': ['Completed', 'Cancelled']
};

function assertTransition(current: string, to: string): void {
  const allowed = ALLOWED_TRANSITIONS[current] ?? [];
  if (!allowed.includes(to)) throw new Error(`Transition ${current} -> ${to} is not allowed`);
}

class TournamentLifecycleService {

  async transitionDraftToRegistration(id: number): Promise<any> {
    const entity = await prisma.tournament.findUnique({ where: { id } });
    if (!entity) throw new Error('Tournament not found: ' + id);
    assertTransition((entity as any).status, 'Registration');
    if ((entity as any).name == null) throw new Error('name is required for Draft -> Registration');
    if ((entity as any).startTime == null) throw new Error('start_time is required for Draft -> Registration');
    const updated = await prisma.tournament.update({ where: { id }, data: { status: 'REGISTRATION' as any } });
    return updated;
  }

  async transitionRegistrationToOngoing(id: number): Promise<any> {
    const entity = await prisma.tournament.findUnique({ where: { id } });
    if (!entity) throw new Error('Tournament not found: ' + id);
    assertTransition((entity as any).status, 'Ongoing');
    const updated = await prisma.tournament.update({ where: { id }, data: { status: 'ONGOING' as any } });
    // TODO: entity.start(); // @after
    return updated;
  }

  async transitionRegistrationToCancelled(id: number): Promise<any> {
    const entity = await prisma.tournament.findUnique({ where: { id } });
    if (!entity) throw new Error('Tournament not found: ' + id);
    assertTransition((entity as any).status, 'Cancelled');
    const updated = await prisma.tournament.update({ where: { id }, data: { status: 'CANCELLED' as any } });
    // TODO: entity.cancel(); // @after
    return updated;
  }

  async transitionOngoingToCompleted(id: number): Promise<any> {
    const entity = await prisma.tournament.findUnique({ where: { id } });
    if (!entity) throw new Error('Tournament not found: ' + id);
    assertTransition((entity as any).status, 'Completed');
    const updated = await prisma.tournament.update({ where: { id }, data: { status: 'COMPLETED' as any } });
    // TODO: entity.complete(); // @after
    // TODO: entity.calculatePrizeDistribution(); // @after
    return updated;
  }

  async transitionOngoingToCancelled(id: number): Promise<any> {
    const entity = await prisma.tournament.findUnique({ where: { id } });
    if (!entity) throw new Error('Tournament not found: ' + id);
    assertTransition((entity as any).status, 'Cancelled');
    const updated = await prisma.tournament.update({ where: { id }, data: { status: 'CANCELLED' as any } });
    // TODO: entity.cancel(); // @after
    return updated;
  }

  async transitionCompletedToDraft(id: number): Promise<any> {
    const entity = await prisma.tournament.findUnique({ where: { id } });
    if (!entity) throw new Error('Tournament not found: ' + id);
    throw new Error('Transition Completed -> Draft is not allowed');
  }

  async transitionCancelledToDraft(id: number): Promise<any> {
    const entity = await prisma.tournament.findUnique({ where: { id } });
    if (!entity) throw new Error('Tournament not found: ' + id);
    throw new Error('Transition Cancelled -> Draft is not allowed');
  }
}
const lifecycleService = new TournamentLifecycleService();


router.get('/', async (req, res) => {
  const q = req.query.q as string | undefined;
  const where = q ? { OR: [{ name: { contains: q } }, { description: { contains: q } }] } : undefined;
  const items = await prisma.tournament.findMany(where ? { where } : undefined);
  res.json(items.map(applyProjection));
});

router.post('/', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.name !== undefined) data.name = body.name;
    if (body.description !== undefined) data.description = body.description;
    if (body.status !== undefined) data.status = body.status;
    if (body.format !== undefined) data.format = body.format;
    if (body.tournamentType !== undefined) data.tournamentType = body.tournamentType;
    if (body.maxPlayers !== undefined) data.maxPlayers = body.maxPlayers;
    if (body.entryFee !== undefined) data.entryFee = body.entryFee;
    if (body.prizePool !== undefined) data.prizePool = body.prizePool;
    if (body.startTime !== undefined) data.startTime = body.startTime != null ? new Date(body.startTime) : null;
    if (body.endTime !== undefined) data.endTime = body.endTime != null ? new Date(body.endTime) : null;
    if (body.isOnline !== undefined) data.isOnline = body.isOnline;
    if (body.location !== undefined) data.location = body.location;
    if (body.rulesText !== undefined) data.rulesText = body.rulesText;
    if (body.createdAt !== undefined) data.createdAt = body.createdAt != null ? new Date(body.createdAt) : null;
    if (body.seasonId !== undefined) data.seasonId = body.seasonId;
    if (body.organizerId !== undefined) data.organizerId = body.organizerId;
  try {
  validate(data);
    const entity = await prisma.tournament.create({ data });
    res.status(201).json(applyProjection(entity));
  } catch (err: any) {
    const status = err?.code === 'P2002' ? 422 : 400;
    res.status(status).json({ error: err?.code === 'P2002' ? 'Value must be unique' : (err?.message ?? 'Validation error') });
  }
});

router.get('/:id', async (req, res) => {
  const entity = await prisma.tournament.findUnique({ where: { id: Number(req.params.id) } });
  if (!entity) return res.status(404).json({ error: 'Not found' });
  res.json(applyProjection(entity));
});

router.put('/:id', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.name !== undefined) data.name = body.name;
    if (body.description !== undefined) data.description = body.description;
    if (body.status !== undefined) data.status = body.status;
    if (body.format !== undefined) data.format = body.format;
    if (body.tournamentType !== undefined) data.tournamentType = body.tournamentType;
    if (body.maxPlayers !== undefined) data.maxPlayers = body.maxPlayers;
    if (body.entryFee !== undefined) data.entryFee = body.entryFee;
    if (body.prizePool !== undefined) data.prizePool = body.prizePool;
    if (body.startTime !== undefined) data.startTime = body.startTime != null ? new Date(body.startTime) : null;
    if (body.endTime !== undefined) data.endTime = body.endTime != null ? new Date(body.endTime) : null;
    if (body.isOnline !== undefined) data.isOnline = body.isOnline;
    if (body.location !== undefined) data.location = body.location;
    if (body.rulesText !== undefined) data.rulesText = body.rulesText;
    if (body.createdAt !== undefined) data.createdAt = body.createdAt != null ? new Date(body.createdAt) : null;
    if (body.seasonId !== undefined) data.seasonId = body.seasonId;
    if (body.organizerId !== undefined) data.organizerId = body.organizerId;
  try {
  validate(data);
    const entity = await prisma.tournament.update({ where: { id: Number(req.params.id) }, data });
    res.json(applyProjection(entity));
  } catch (err: any) {
    const status = err?.code === 'P2025' ? 404 : err?.code === 'P2002' ? 422 : 400;
    res.status(status).json({ error: err?.code === 'P2002' ? 'Value must be unique' : (err?.message ?? 'Error') });
  }
});

router.patch('/:id', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.name !== undefined) data.name = body.name;
    if (body.description !== undefined) data.description = body.description;
    if (body.status !== undefined) data.status = body.status;
    if (body.format !== undefined) data.format = body.format;
    if (body.tournamentType !== undefined) data.tournamentType = body.tournamentType;
    if (body.maxPlayers !== undefined) data.maxPlayers = body.maxPlayers;
    if (body.entryFee !== undefined) data.entryFee = body.entryFee;
    if (body.prizePool !== undefined) data.prizePool = body.prizePool;
    if (body.startTime !== undefined) data.startTime = body.startTime != null ? new Date(body.startTime) : null;
    if (body.endTime !== undefined) data.endTime = body.endTime != null ? new Date(body.endTime) : null;
    if (body.isOnline !== undefined) data.isOnline = body.isOnline;
    if (body.location !== undefined) data.location = body.location;
    if (body.rulesText !== undefined) data.rulesText = body.rulesText;
    if (body.createdAt !== undefined) data.createdAt = body.createdAt != null ? new Date(body.createdAt) : null;
    if (body.seasonId !== undefined) data.seasonId = body.seasonId;
    if (body.organizerId !== undefined) data.organizerId = body.organizerId;
  try {
  validate(data);
    const entity = await prisma.tournament.update({ where: { id: Number(req.params.id) }, data });
    res.json(applyProjection(entity));
  } catch (err: any) {
    const status = err?.code === 'P2025' ? 404 : err?.code === 'P2002' ? 422 : 400;
    res.status(status).json({ error: err?.code === 'P2002' ? 'Value must be unique' : (err?.message ?? 'Error') });
  }
});

router.post('/:id/start', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.start(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/cancel', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.cancel(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/complete', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.complete(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/rounds', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.generate_round(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.get('/:id/prizes', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    const result = await service.calculate_prize_distribution(id);
    res.json({ result });
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/register', async (req, res) => {
  const id = Number((req.params as any).id);
  const playerId = req.body.playerId;
  const deckId = req.body.deckId;
  try {
    await service.register_player(id, playerId, deckId);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.get('/:id/full', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    const result = await service.is_full(id);
    res.json({ result });
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.patch('/:id/transitions/draft-to-registration', async (req, res) => {
  const id = Number(req.params.id);
  try {
    const entity = await lifecycleService.transitionDraftToRegistration(id);
    res.json(entity);
  } catch (err: any) {
    const status = err?.message?.includes('not allowed') || err?.message?.includes('is not allowed') ? 409
      : err?.message?.includes('required') || err?.message?.includes('must be') ? 422
      : 404;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id/transitions/registration-to-ongoing', async (req, res) => {
  const id = Number(req.params.id);
  try {
    const entity = await lifecycleService.transitionRegistrationToOngoing(id);
    res.json(entity);
  } catch (err: any) {
    const status = err?.message?.includes('not allowed') || err?.message?.includes('is not allowed') ? 409
      : err?.message?.includes('required') || err?.message?.includes('must be') ? 422
      : 404;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id/transitions/registration-to-cancelled', async (req, res) => {
  const id = Number(req.params.id);
  try {
    const entity = await lifecycleService.transitionRegistrationToCancelled(id);
    res.json(entity);
  } catch (err: any) {
    const status = err?.message?.includes('not allowed') || err?.message?.includes('is not allowed') ? 409
      : err?.message?.includes('required') || err?.message?.includes('must be') ? 422
      : 404;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id/transitions/ongoing-to-completed', async (req, res) => {
  const id = Number(req.params.id);
  try {
    const entity = await lifecycleService.transitionOngoingToCompleted(id);
    res.json(entity);
  } catch (err: any) {
    const status = err?.message?.includes('not allowed') || err?.message?.includes('is not allowed') ? 409
      : err?.message?.includes('required') || err?.message?.includes('must be') ? 422
      : 404;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id/transitions/ongoing-to-cancelled', async (req, res) => {
  const id = Number(req.params.id);
  try {
    const entity = await lifecycleService.transitionOngoingToCancelled(id);
    res.json(entity);
  } catch (err: any) {
    const status = err?.message?.includes('not allowed') || err?.message?.includes('is not allowed') ? 409
      : err?.message?.includes('required') || err?.message?.includes('must be') ? 422
      : 404;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id/transitions/completed-to-draft', async (req, res) => {
  const id = Number(req.params.id);
  try {
    const entity = await lifecycleService.transitionCompletedToDraft(id);
    res.json(entity);
  } catch (err: any) {
    const status = err?.message?.includes('not allowed') || err?.message?.includes('is not allowed') ? 409
      : err?.message?.includes('required') || err?.message?.includes('must be') ? 422
      : 404;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id/transitions/cancelled-to-draft', async (req, res) => {
  const id = Number(req.params.id);
  try {
    const entity = await lifecycleService.transitionCancelledToDraft(id);
    res.json(entity);
  } catch (err: any) {
    const status = err?.message?.includes('not allowed') || err?.message?.includes('is not allowed') ? 409
      : err?.message?.includes('required') || err?.message?.includes('must be') ? 422
      : 404;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});
export default router;

File: gin/internal/handler/tournaments/tournaments.go
package handler_tournaments

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	model "cards_project/internal/model/tournaments"
	"cards_project/internal/handler"
)

type TournamentHandler struct { db *gorm.DB }

func NewTournamentHandler(db *gorm.DB) *TournamentHandler {
	return &TournamentHandler{db: db}
}

func (h *TournamentHandler) RegisterRoutes(r gin.IRouter) {
	g := r.Group("/api/tournaments")
	g.GET("", h.List)
	g.POST("", h.Create)
	g.GET("/:id", h.Get)
	g.PUT("/:id", h.Update)
	g.PATCH("/:id", h.Patch)
	g.DELETE("/:id", h.Delete)
	g.POST("/:id/start", h.Start)
	g.POST("/:id/cancel", h.Cancel)
	g.POST("/:id/complete", h.Complete)
	g.POST("/:id/rounds", h.GenerateRound)
	g.GET("/:id/prizes", h.CalculatePrizeDistribution)
}

func (h *TournamentHandler) List(c *gin.Context) {
	skip, limit := handler.Paginate(c)
	var rows []model.Tournament
	if err := h.db.Offset(skip).Limit(limit).Find(&rows).Error; err != nil {
		handler.DbError(c, err); return
	}
	out := make([]model.TournamentResponse, len(rows))
	for i := range rows { out[i] = rows[i].ToResponse() }
	c.JSON(http.StatusOK, out)
}

func (h *TournamentHandler) Create(c *gin.Context) {
	var req model.TournamentCreateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	if msgs := validateTournament(&req); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	row := model.Tournament{}
	row.Name = req.Name
	row.Description = req.Description
	row.Format = req.Format
	row.TournamentType = req.TournamentType
	row.Status = req.Status
	row.MaxPlayers = req.MaxPlayers
	row.EntryFee = req.EntryFee
	row.PrizePool = req.PrizePool
	row.StartTime = req.StartTime
	row.EndTime = req.EndTime
	row.IsOnline = req.IsOnline
	row.Location = req.Location
	row.RulesText = req.RulesText
	row.SeasonID = req.SeasonID
	row.OrganizerID = req.OrganizerID
	if err := h.db.Create(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusCreated, row.ToResponse())
}

func (h *TournamentHandler) Get(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Tournament
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Tournament"); return }
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *TournamentHandler) Update(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Tournament
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Tournament"); return }
		handler.DbError(c, err); return
	}
	var req model.TournamentUpdateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	row.ApplyUpdate(req)
	createReq := toCreateRequestTournament(&row)
	if msgs := validateTournament(&createReq); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	if err := h.db.Save(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *TournamentHandler) Patch(c *gin.Context) { h.Update(c) }

func (h *TournamentHandler) Delete(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	if err := h.db.Delete(&model.Tournament{}, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Tournament"); return }
		handler.DbError(c, err); return
	}
	c.Status(http.StatusNoContent)
}

func (h *TournamentHandler) Start(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Tournament
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Tournament"); return }
		handler.DbError(c, err); return
	}
	err := row.Start()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *TournamentHandler) Cancel(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Tournament
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Tournament"); return }
		handler.DbError(c, err); return
	}
	err := row.Cancel()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *TournamentHandler) Complete(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Tournament
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Tournament"); return }
		handler.DbError(c, err); return
	}
	err := row.Complete()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *TournamentHandler) GenerateRound(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Tournament
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Tournament"); return }
		handler.DbError(c, err); return
	}
	err := row.GenerateRound()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *TournamentHandler) CalculatePrizeDistribution(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Tournament
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Tournament"); return }
		handler.DbError(c, err); return
	}
	result, err := row.CalculatePrizeDistribution()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.JSON(http.StatusOK, gin.H{"result": result})
}

func validateTournament(req *model.TournamentCreateRequest) []string {
	var errs []string
	if !((req.MaxPlayers >= 2 && req.MaxPlayers <= 512)) {
		errs = append(errs, "Tournament must allow between 2 and 512 players")
	}
	if !(float64(req.EntryFee) >= 0) {
		errs = append(errs, "Entry fee must not be negative")
	}
	if !(float64(req.PrizePool) >= 0) {
		errs = append(errs, "Prize pool must not be negative")
	}
	if !((!( req.EndTime != nil ) || (*req.EndTime > req.StartTime))) {
		errs = append(errs, "End time must be after start time")
	}
	return errs
}

func toCreateRequestTournament(m *model.Tournament) model.TournamentCreateRequest {
	return model.TournamentCreateRequest{
		Name: m.Name,
		Description: m.Description,
		Format: m.Format,
		TournamentType: m.TournamentType,
		Status: m.Status,
		MaxPlayers: m.MaxPlayers,
		EntryFee: m.EntryFee,
		PrizePool: m.PrizePool,
		StartTime: m.StartTime,
		EndTime: m.EndTime,
		IsOnline: m.IsOnline,
		Location: m.Location,
		RulesText: m.RulesText,
		SeasonID: m.SeasonID,
		OrganizerID: m.OrganizerID,
	}
}
